esp/clients/wsdfuaccess/wsdfuaccess.cpp

```diff
@@ -59,21 +89,667 @@ bool getFileAccess(StringBuffer &metaInfo, const char *serviceUrl, const char *j
     lfn.getCluster(cluster);
     lfn.get(lfnName); // remove cluster if present
 
+    IEspDFUFileAccessRequestBase &requestBase = dfuReq->updateRequestBase();
     requestBase.setName(lfnName);
     requestBase.setCluster(cluster);
     requestBase.setExpirySeconds(expirySecs);
-    requestBase.setAccessRole(CFileAccessRole_Engine);
-    requestBase.setAccessType(translateToCSecAccessAccessType(access));
-    requestBase.setJobId(jobId);
+    requestBase.setJobId(requestId);
+    requestBase.setAccessRole(CFileAccessRole_External);
+    requestBase.setAccessType(CSecAccessType_Read);
+    requestBase.setReturnBinTypeInfo(true);
+
+    return dfuClient->DFUFileAccess(dfuReq);
+}
+
+static IClientDFUFileAccessResponse *doLookupDFUFile(IClientWsDfu *dfuClient, const char *logicalName, const char *requestId, unsigned expirySecs)
+{
+    Owned<IClientDFULookupFileRequest> dfuReq = dfuClient->createDFULookupFileRequest();
 
-    Owned<IClientDFUFileAccessResponse> dfuResp = dfuClient->DFUFileAccess(dfuReq);
+    CDfsLogicalFileName lfn;
+    lfn.set(logicalName);
+
+    StringBuffer cluster, lfnName;
+    lfn.getCluster(cluster);
+    lfn.get(lfnName); // remove cluster if present
+
+    dfuReq->setName(lfnName);
+    dfuReq->setCluster(cluster);
+    dfuReq->setExpirySeconds(expirySecs);
+    dfuReq->setRequestId(requestId);
+#ifdef TEST_RETURNTEXTRESPONSE
+    dfuReq->setReturnTextResponse(true);
+#endif
+
+    return dfuClient->DFULookupFile(dfuReq);
+}
+
+static IDFUFileAccess *doLookupDFUFileHandleLegacy(const char *serviceUrl, const char *logicalName, const char *requestId, unsigned expirySecs, const char *user, const char *password)
+{
+    Owned<IClientWsDfu> dfuClient = createWsDfuClient();
+    dfuClient->addServiceUrl(serviceUrl);
+    dfuClient->setUsernameToken(user, password, "");
+    Owned<IClientDFUFileAccessResponse> dfuResp;
+    try
+    {
+        dfuResp.setown(doLookupDFUFile(dfuClient, logicalName, requestId, expirySecs));
+    }
+    catch (IException *e)
+    {
+        /* NB: there should really be a different IException class and a specific error code
+         * The server knows it's an unsupported method.
+         */
+        if (SOAP_SERVER_ERROR != e->errorCode())
```

Comment:
It might be less expensive to explicitly check the target HPCC version to know which wsdfu to target.

Reply:
It would be better yes, but further to our offline discussion last week:

Esp should offer a standard way to discover a service version, but there is none at the moment.
I'm going to open a JIRA for this.

You mentioned a kludgy route that's used sometimes involving a soap call and parsing soap:address.
As a stop-gap that may be worth pursuing.

@rpastrana - can you point to a code example of this? Is it in a utility class/method I can reuse?

Reply:
@jakesmith no example on hand, I don't think this situation has come up before to be honest (deprecated methods, in favor of new methods). 

Until we get a sensible mechanism from the ESP to provide the interface version we might be stuck with your approach, the stop-gap measures I mentioned are indeed a kludge and not necessarily endorsed:

1) fetch HPCC build ver from wssmc and infer wsdfu interface ver.
http://n.n.n.n:8010/WsSMC/Activity returns a "Build" field, which would need to be parsed.

2) Requesting the WSDL http://n.n.n.n:8010/WsDfu/?wsdl
Will provide the default/latest interface version, but also will list the supported methods (but forces the client to interpret WSDL)

3) Requesting esp form: http://n.n.n.n:8010/WsDfu/ <- esp form returns the following HTML:
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"><title>Enterprise Services Platform</title></head><body><h1>WsDfu</h1><h2>**Version: 1.41**, Build: internal_7.3.0-trunk0[remotes/origin/master-0-g5427b5]</h2>

Reply:
it still might be worth kludge, but it's not critical just an improvement for connecting to 7.0.x esp's.
Let's concentrate on getting this merged and I'll open a new JIRA re. adding a kludged getEspVersion() to use in the interim until HPCC-21653 is available